src/utils/sort.hpp

```diff
@@ -31,6 +31,28 @@
 
 namespace parthenon {
 
+// Returns the upper bound (or the array size if value has not been found)
+// Could/Should be replaced with a Kokkos std version once available (currently schedule
+// for 4.2 release).
+template <class T>
+KOKKOS_INLINE_FUNCTION int upper_bound(const T &arr, Real val) {
+  int l = 0;
+  int r = arr.extent_int(0);
+  int m;
+  while (l < r) {
+    m = l + (r - l) / 2;
+    if (val >= arr(m)) {
+      l = m + 1;
+    } else {
+      r = m;
+    }
+  }
+  if (l < arr.extent_int(0) && val >= arr(l)) {
+    l++;
+  }
+  return l;
+}
```

Comment:
I'm not sure I follow.
This is an inline function that is called by each thread, does not modify any data, and returns a single value (at a dedicated memory location).
Where would that cause issues?

Reply:
Doesn't this function only work if you call it on either 1 array per thread? The way you're calling it, you launch a single thread on the array. That's all I mean. If someone is expecting this to be some kind of `kokkos_scan` based function that leverages multiple threads, that isn't the case.

Reply:
The assumption here is that `arr` is not modified while the index/lower bound `l` is determined, so yes, when you use this function side a parallel scan where the `arr` is the result of the scan, then this will cause issue.
Though I'm not sure this (finding and index/doing a binary search in a changing array/container) is even legal for the `std::upper_bound`.
Or maybe I'm still not understanding the issue.

Reply:
There's nothing technically wrong with the function, so I think my issue is more conceptual than anything. This function is exactly what you want for the histogram code, I'm just worried about it being misused since it lives in `utils`.

Reply:
I put it specifically in `utils` so that it can be reused as drop in replacement for `https://en.cppreference.com/w/cpp/algorithm/upper_bound` on device until a Kokkos version is available.
So there should be no extra gotchas (as far as I'm aware of) in comparison to `std::upper_bound`.

Reply:
If not already present, can you add a comment that it has exactly the same API as `std::upper_bound` but on device? I think that will disambiguate for me

Reply:
Added a comment (and highlighted the differences).